crates/dojo-world/src/manifest.rs

```diff
@@ -106,133 +83,283 @@ pub struct ComputedValueEntrypoint {
 
 #[serde_as]
 #[derive(Clone, Default, Debug, Serialize, Deserialize, PartialEq)]
-pub struct Contract {
-    pub name: SmolStr,
-    #[serde_as(as = "Option<UfeHex>")]
-    pub address: Option<FieldElement>,
+#[serde(tag = "kind")]
+pub struct DojoContract {
     #[serde_as(as = "UfeHex")]
     pub class_hash: FieldElement,
-    pub abi: Option<abi::Contract>,
+    pub abi: Option<String>,
+    #[serde_as(as = "Option<UfeHex>")]
+    pub address: Option<FieldElement>,
     pub reads: Vec<String>,
     pub writes: Vec<String>,
     pub computed: Vec<ComputedValueEntrypoint>,
 }
 
+impl ManifestMethods for DojoContract {
+    fn abi(&self) -> Option<&String> {
+        self.abi.as_ref()
+    }
+
+    fn set_abi(&mut self, abi: Option<String>) {
+        self.abi = abi;
+    }
+
+    fn class_hash(&self) -> &FieldElement {
+        self.class_hash.as_ref()
+    }
+
+    fn set_class_hash(&mut self, class_hash: FieldElement) {
+        self.class_hash = class_hash;
+    }
+}
```

Comment:
The generic approach looks promising, the impl of `ManifestMethods` seems repetitive at first, but I don't think there's a better approach to factorize that. This may also evolve independently for any inner type.

Reply:
we could use macro reduce code duplication but i kept it simple for review, maybe before merging we can do that.